### ISSUE TITLE
Potential problem with actor handles

### ISSUE THREAD

**Issue by pcmoritz** [ray-project/ray#3056]
The following crashes XRay and makes legacy ray hang:

```python
import ray

ray.init(use_raylet=True)

@ray.remote
class Actor(object):
    def f(self):
        return 1

@ray.remote
class Bctor(object):
    def __init__(self, constructor):
        self.actor = constructor()

    def step(self):
        ray.get(self.actor.f.remote())

a = Actor.remote()

b = Bctor.remote(lambda: a)

ray.get(b.step.remote())
ray.get(b.step.remote())
```

**Comment by robertnishihara** [comment]
Can you explain what the issue is? This looks like it should work.

**Comment by pcmoritz** [comment]
The symptom is this:

```
F1014 12:59:52.206046 2571563904 node_manager.cc:72]  Check failed: spec.ActorCounter() == expected_task_counter Expected actor counter: 0, got: 1
*** Check failure stack trace: ***
    @        0x10a4cc82d  google::LogMessage::Fail()
    @        0x10a4ca64e  google::LogMessage::SendToLog()
    @        0x10a4cb4cf  google::LogMessage::Flush()
    @        0x10a4cb309  google::LogMessage::~LogMessage()
    @        0x10a4cb5c5  google::LogMessage::~LogMessage()
    @        0x10a42c3dd  ray::RayLog::~RayLog()
    @        0x10a47b417  ray::raylet::NodeManager::AssignTask()
    @        0x10a47ab08  ray::raylet::NodeManager::DispatchTasks()
    @        0x10a4828b1  ray::raylet::NodeManager::EnqueuePlaceableTask()
    @        0x10a479abc  ray::raylet::NodeManager::SubmitTask()
    @        0x10a47eaef  ray::raylet::NodeManager::ProcessSubmitTaskMessage()
    @        0x10a47cbd9  ray::raylet::NodeManager::ProcessClientMessage()
    @        0x10a3de63c  std::__1::__function::__func<>::operator()()
    @        0x10a42d853  ray::ClientConnection<>::ProcessMessage()
    @        0x10a430db8  boost::asio::detail::reactive_socket_recv_op<>::do_complete()
    @        0x10a3d05b8  boost::asio::detail::task_io_service::do_run_one()
    @        0x10a3d0121  boost::asio::detail::task_io_service::run()
    @        0x10a3cb7de  main
```

New actor handle IDs are being created that are not registered in the backend. One possibility is that every time we call compute_actor_handle_id_non_forked, the actor counter needs to be reset too. But not sure if that's the right fix.

**Comment by stephanie-wang** [comment]
Hmm, I'm not really sure why this is happening since it looks like the actor counter should get reset [here](https://github.com/ray-project/ray/blob/master/python/ray/actor.py#L662). Can you check to make sure that that line is getting hit?

**Comment by pcmoritz** [comment]
It does get reset, but only when the actor handle is unpickled. However every time a method on the unpickled actor handle is called, a new actor handle is created by compute_actor_handle_id_non_forked and for that, the counter is not reset. See here: https://github.com/ray-project/ray/blob/master/python/ray/actor.py#L558

It seems to me very dangerous to create a new actor handle id every time we call an actor handle. This can easily lead to blowup of datastructures in the raylet.

Maybe actor_handle_id was supposed to be assigned to self._ray_actor_handle_id in the line above?

**Comment by stephanie-wang** [comment]
That seems like a bug; the actor handle should only get created once, when the handle is first unpickled. I believe that we should be setting `self._ray_actor_handle_id` earlier, somewhere around [here](https://github.com/ray-project/ray/blob/master/python/ray/actor.py#L697). I don't really understand the use of `ray_forking` in that code, though.

@robertnishihara, it looks like you wrote that commit. Can you explain what happens when `ray_forking = False` and why it is necessary?

**Comment by robertnishihara** [comment]
Creating a new actor handle was intentional. There are a couple cases:

1. We pass an actor handle into a task/actor. This is the `forking=True` case. It creates a new actor handle for the task/actor. In this case we generate the new actor handle ID deterministically.

2. A remote function or actor class includes an actor handle in its closure (but the actor handle is not passed in). When the remote function or actor class is unpickled, we create an "uninitialized" actor handle. Then when a task actually tries to use the actor handle, we generate a new actor handle ID for it (deterministically based on the actor task ID). Note that we can't generate the actor handle ID deterministically when the actor handle is unpickled because it can be unpickled on multiple workers.

If we want to generate actor handle IDs deterministically, then we need a separate one per task (or per actor) that uses the actor handle (we currently also do one per actor method to simplify the code, so we could reduce the redundancy a bit in the actor case).

3. There's a third case where someone just pickles an actor handle and then unpickles it somewhere else (e.g., a different driver). This is the same code path as case 2.

@pcmoritz is right that there is a performance issue here with the explosion of actor handles. I may be overlooking some simple solution. Ideally we'd just find a way to make actor handles super lightweight.

**Comment by robertnishihara** [comment]
One possibility might be to have pickled actor handles essentially not have any dependencies with other actor handles, they only maintain a local count of methods that have been invoked and the count is reset every time a new task starts.

Also, one thing I wasn't sure how to handle (and which currently breaks) is what to do when an actor handle is unpickled multiple times in the same task.

**Comment by robertnishihara** [comment]
Fixed by #3074.
